[LETS-527] Deal the NULL previous version LSA with TOO_NEW_FOR_SNAPSHOT

wrlawodms opened the pull request:
http://jira.cubrid.org/browse/LETS-527

I missed a normal case where the previous version can be NULL while the snapshot result is TOO_NEW_FOR_SNAPSHOT.
It means after a snapshot is taken, a record is newly inserted.

wrlawodms committed:
fix to return when null prev lsa

cristiarg reviewed (approved):
It would be nice to add an extensive comment before `return S_DOESNT..` to explain the scenario. At your choice. Thanks.

wrlawodms committed:
const

wrlawodms committed:
Keep the assertion, and filter the case in the caller

wrlawodms commented:
@cristiarg I have added a comment for the case. On second thought, it seems awkward to get a previous version with the NULL previous LSA, which even always fails and returns S_DOESNT_EXIST. So, I've changed to keep the assertion and filter the case in the caller. 

Plus, I changed the description for the case. The case seems only when a record is newly inserted after the snapshot is taken because the ins_mvccid and prev_version_lsa are cleared together when every active transaction can see it, and in turn the snapshot can't be TOO_NEW_FOR_SNAPSHOT.

wrlawodms committed:
trivials

cristiarg commented on src/storage/heap_file.c:
```diff
@@ -25255,10 +25255,18 @@ heap_get_visible_version_internal (THREAD_ENTRY * thread_p, HEAP_GET_CONTEXT * c
       snapshot_res = mvcc_snapshot->snapshot_fnc (thread_p, &mvcc_header, mvcc_snapshot);
       if (snapshot_res == TOO_NEW_FOR_SNAPSHOT)
 	{
+	  const LOG_LSA *prev_lsa = &MVCC_GET_PREV_VERSION_LSA (&mvcc_header);
```
Nit:
```suggestion
	  const LOG_LSA *const prev_lsa = &MVCC_GET_PREV_VERSION_LSA (&mvcc_header);
```

hornetmj commented on src/storage/heap_file.c:
```diff
@@ -25255,10 +25255,18 @@ heap_get_visible_version_internal (THREAD_ENTRY * thread_p, HEAP_GET_CONTEXT * c
+	  if (prev_lsa->is_null ())
+	    {
+	      /* The record is newly inserted after the snapshot is taken. */
+	      scan = S_DOESNT_EXIST;
```
Is there any reason to use the S_DOESNT_EXIST iso the S_SNAPSHOT_NOT_SATISFIED?

wrlawodms replied:
It is what was returned before in the case.

hornetmj commented on src/storage/heap_file.c:
```diff
@@ -25255,10 +25255,18 @@ heap_get_visible_version_internal (THREAD_ENTRY * thread_p, HEAP_GET_CONTEXT * c
       if (snapshot_res == TOO_NEW_FOR_SNAPSHOT)
 	{
+	  const LOG_LSA *prev_lsa = &MVCC_GET_PREV_VERSION_LSA (&mvcc_header);
+	  if (prev_lsa->is_null ())
```
In develop branch, the heap_get_visible_version_from_log() can handle when the prev_lsa is NULL. But, this function cannot handle it in the scalability_dev because assert (!previous_version_lsa->is_null ()) condition. If we convince the assert condtion, then it is better not to make this situation when we build the recdes. Isn't it?

hornetmj replied:
But, I think we cannot because the REC_BIGONE type. So, how about remove the assert() and add the if (prev_lsa->is_null ()) condition in the heap_get_visible_version_from_log() function.

cristiarg replied:
@hornetmj @wrlawodms 
Please be aware that there is also a related issue which is a regression coming from QA - http://jira.cubrid.org/browse/LETS-532 - and the test bumped into the assert. So the assert must be evaluated also in the context of this regression (and, possibly others) in the most conservative way.
Me, personally, I would leave the assert if it does not offend any other tests - with the idea that it's best to contain the NULL LSA scenarios as much as possible (ie: be notified if other tests bump into it). If the assert fails again in a legitimate scenario, we can re-evaluate.

joohok replied:
If the `previous_version_lsa` is null inside the `heap_get_visible_version_from_log()` function, it does nothing(Line 25000) and throws an error (S_DOESNT_EXIST). Rather than adding asserts in `heap_get_visible_version_from_log()`, how about adding exception handling earlier?

wrlawodms replied:
@hornetmj 
>In develop branch, the heap_get_visible_version_from_log() can handle when the prev_lsa is NULL.

I don't think it was handled correctly because it used the NULL previous LSA to compare with:
```if (LSA_LT (&oldest_prior_lsa, previous_version_lsa))```
, which looks unreasonable and triggers unnecessary appending prior lists.

>But, I think we cannot because the REC_BIGONE type.

The big one record can have the NULL previous LSA, and other types also can be considered to have the NULL lsa here (if the flag is off). However, it's only the case when the record is newly inserted. In this case, it seems unreasonable to find the previous version. On top of that, the semantics at all seems awkward: "to find a previous version in the LOG with NULL LSA (N/A location)".

wrlawodms replied:
@joohok You mean the exception handler inside the function like hornetmj's opinion?

wrlawodms replied:
@hornetmj oh sorry to confuse you. It doesn't trigger.
> which looks unreasonable and causes unnecessary appending prior lists.

joohok replied:
> On top of that, the semantics at all seems awkward to find a previous version in the LOG with NULL LSA (N/A location);

I think this is right. Now I agree with your change. I think exception handler is unnecessary.

hornetmj replied:
@wrlawodms @joohok I understand what you mean. However, in my point of view, it is better to handle in the heap_get_visible_version_from_log() from the heap_* API user's point. From a semantic point of view, I think there is no problem. However, I will follow the comments of the two reviewers.

wrlawodms committed:
const
